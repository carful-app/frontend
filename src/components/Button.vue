<script setup lang="ts">
const props = defineProps<{
  color: 'light' | 'dark' | 'blue'
  text?: boolean
  outline?: boolean
  loading?: boolean
  btnClasses?: string
}>()

const { color, text, outline, btnClasses } = toRefs(props)

const externalBtnClasses = computed(() => {
  return btnClasses?.value ? btnClasses.value.split(' ').reduce((acc, curr) => ({ ...acc, [curr]: true }), {}) : []
})

interface ButtonClasses {
  btn: boolean
  'btn-light'?: boolean
  'btn-dark'?: boolean
  'btn-blue'?: boolean
  'btn-outline-light'?: boolean
  'btn-outline-dark'?: boolean
  'btn-outline-blue'?: boolean
  'btn-text-light'?: boolean
  'btn-text-dark'?: boolean
  'btn-text-blue'?: boolean
}

const internalBtnClasses = computed(() => {
  const classes: ButtonClasses = {
    btn: true,
  }

  if (text?.value) {
    classes[`btn-text-${color.value}`] = true
  } else if (outline?.value) {
    classes[`btn-outline-${color.value}`] = true
  } else {
    classes[`btn-${color.value}`] = true
  }

  return classes
})

defineEmits(['click'])
</script>

<template>
  <div class="d-grid">
    <button :class="{ ...internalBtnClasses, ...externalBtnClasses }" @click="$emit('click')" :disabled="loading">
      <span class="spinner-border spinner-border-sm" role="status" aria-hidden="true" v-if="loading"></span>
      <slot />
    </button>
  </div>
</template>

<style lang="sass" scoped>
@import '@/assets/styles/variables.sass'
@import '~bootstrap/scss/bootstrap-utilities'

.btn-light
  @include button-variant($color-light-blue, $color-light-blue, $color-white)

.btn-dark
  @include button-variant($color-dark-blue, $color-dark-blue, $color-white)

.btn-blue
  @include button-variant($color-blue, $color-blue, $color-white)

.btn-outline-light
  @include button-outline-variant($color-light-blue)

.btn-outline-dark
  @include button-outline-variant($color-dark-blue)

.btn-outline-blue
  @include button-outline-variant($color-blue)

.btn-text-light
  @include button-outline-variant($color-light-blue)
  border: none

.btn-text-dark
  @include button-outline-variant($color-dark-blue)
  border: none

.btn-text-blue
  @include button-outline-variant($color-blue)
  border: none
</style>
