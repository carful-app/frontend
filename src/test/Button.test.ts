import { mount } from '@vue/test-utils'
import Button from '@/components/common/Button.vue'

describe('test Button component', () => {
  // test if the component recieves props as expected
  test('props as expected', () => {
    const color = 'light'
    const cmp = mount(Button, {
      props: {
        color,
      },
    })
    expect(cmp.props()).toEqual({
      color,
      text: false,
      outline: false,
      loading: false,
      isDisabled: false,
      btnClasses: undefined,
    })
  })

  // test if the component renders correctly
  test('renders correctly', () => {
    const cmp = mount(Button, {
      props: {
        color: 'light',
      },
      slots: {
        default: 'Button',
      },
    })
    expect(cmp.element).toMatchSnapshot()
  })

  // test if the component emits events as expected
  test('emits events as expected', async () => {
    const cmp = mount(Button, {
      props: {
        color: 'light',
      },
      slots: {
        default: 'Button',
      },
    })
    await cmp.find('button').trigger('click')
    expect(cmp.emitted().click).toBeTruthy()
  })

  // test if the component has the correct classes
  test('has the correct classes', () => {
    const cmp = mount(Button, {
      props: {
        color: 'light',
      },
      slots: {
        default: 'Button',
      },
    })
    const btn = cmp.find('button')
    expect(btn.classes()).toContain('btn')
    expect(btn.classes()).toContain('btn-light')
  })

  // test if the component has the correct classes when props change
  test('has the correct classes when props change', async () => {
    const cmp = mount(Button, {
      props: {
        color: 'light',
      },
      slots: {
        default: 'Button',
      },
    })
    const btn = cmp.find('button')
    expect(btn.classes()).toContain('btn')
    expect(btn.classes()).toContain('btn-light')
    await cmp.setProps({ color: 'dark' })
    expect(btn.classes()).toContain('btn-dark')
  })

  //test if the component is disabled
  test('is disabled', async () => {
    const cmp = mount(Button, {
      props: {
        color: 'light',
        isDisabled: true,
      },
      slots: {
        default: 'Button',
      },
    })
    expect(cmp.find('button').attributes()).toContain({ disabled: '' })
  })

  // test if the component is loading and disabled
  test('is loading and disabled', async () => {
    const cmp = mount(Button, {
      props: {
        color: 'light',
        loading: true,
      },
      slots: {
        default: 'Button',
      },
    })
    expect(cmp.find('span.spinner-border').exists()).toBeTruthy()
    expect(cmp.find('button').attributes()).toContain({ disabled: '' })
  })

  // test all variants of the component
  test('all variants', async () => {
    const cmp = mount(Button, {
      props: {
        color: 'light',
      },
      slots: {
        default: 'Button',
      },
    })
    expect(cmp.find('button').classes()).toContain('btn-light')

    await cmp.setProps({ color: 'dark' })
    expect(cmp.find('button').classes()).toContain('btn-dark')

    await cmp.setProps({ color: 'blue' })
    expect(cmp.find('button').classes()).toContain('btn-blue')

    // text variants
    await cmp.setProps({ color: 'light', text: true })
    expect(cmp.find('button').classes()).toContain('btn-text-light')

    await cmp.setProps({ color: 'dark' })
    expect(cmp.find('button').classes()).toContain('btn-text-dark')

    await cmp.setProps({ color: 'blue' })
    expect(cmp.find('button').classes()).toContain('btn-text-blue')

    // outline variants
    await cmp.setProps({ color: 'light', outline: true, text: false })
    expect(cmp.find('button').classes()).toContain('btn-outline-light')

    await cmp.setProps({ color: 'dark' })
    expect(cmp.find('button').classes()).toContain('btn-outline-dark')

    await cmp.setProps({ color: 'blue' })
    expect(cmp.find('button').classes()).toContain('btn-outline-blue')
  })

  // test if the component has the correct classes when btnClasses prop is passed
  test('has the correct classes when btnClasses prop is passed', async () => {
    const cmp = mount(Button, {
      props: {
        color: 'light',
        btnClasses: 'btn-lg',
      },
      slots: {
        default: 'Button',
      },
    })
    const btn = cmp.find('button')
    expect(btn.classes()).toContain('btn')
    expect(btn.classes()).toContain('btn-light')
    expect(btn.classes()).toContain('btn-lg')
  })
})
