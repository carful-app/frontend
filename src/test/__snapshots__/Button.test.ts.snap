// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test Button component > renders correctly 1`] = `
<div
  class="d-grid"
  data-v-762a4901=""
>
  <button
    class="btn btn-light"
    data-v-762a4901=""
  >
    <!--v-if-->
    
    Button
    
  </button>
</div>
`;
